napari/_qt/qt_viewer.py

```diff
@@ -289,26 +289,90 @@ def _toggle_console(self):
         self.buttons.consoleButton.style().unpolish(self.buttons.consoleButton)
         self.buttons.consoleButton.style().polish(self.buttons.consoleButton)
 
-    def on_mouse_move(self, event):
-        """Called whenever mouse moves over canvas.
+    def on_mouse_press(self, event):
+        """Called whenever mouse pressed in canvas.
         """
+        if event.pos is None:
+            return
+
+        event = ReadOnlyWrapper(event)
+        for mouse_drag_func in self.viewer.mouse_drag_callbacks:
+            gen = mouse_drag_func(self.viewer, event)
+            if inspect.isgeneratorfunction(mouse_drag_func):
+                try:
+                    next(gen)
+                    self.viewer._mouse_drag_gen[mouse_drag_func] = gen
+                    self.viewer._persisted_mouse_event[gen] = event
+                except StopIteration:
+                    pass
+
         layer = self.viewer.active_layer
         if layer is not None:
-            self.layer_to_visual[layer].on_mouse_move(event)
+            self.layer_to_visual[layer].on_mouse_press(event)
+            for mouse_drag_func in layer.mouse_drag_callbacks:
+                gen = mouse_drag_func(layer, event)
+                if inspect.isgeneratorfunction(mouse_drag_func):
+                    try:
+                        next(gen)
+                        layer._mouse_drag_gen[mouse_drag_func] = gen
+                        layer._persisted_mouse_event[gen] = event
```

Comment:
This code is repeated in all the `on_mouse_*` methods... Would it be possible to unify it somewhere?

Reply:
possibly, i'd be very open to suggestions. maybe can come in a refactor

Reply:
unified in `utils.misc`